R/crunch-data-frame.R

```diff
@@ -0,0 +1,292 @@
+    stopifnot(is.dataset(dataset))
+    
+    if (include.hidden) {
+        var_names <- aliases(allVariables(dataset))
+    } else {
+        var_names <- aliases(variables(dataset))
+    }
+    
+    out <- new.env()
+    attr(out, "crunchDataset") <- dataset
+    attr(out, "col_names") <- var_names
+    attr(out, "mode") <- categorical.mode
+    
+    with(out, {
+        ## Note the difference from as.environment: wrapped in as.vector
+        for (.a in var_names) {
+            eval(substitute(delayedAssign(v, {
+                get_var_from_server(v, ds, mode)
+            }),
+            list(v=.a, ds=out, mode = attr(out, "mode"))))
+        }
+    })
+    
+    # set the order of the dataset based on row.order, if no row.order is 
+    # given return all rows in the dataset in the order they appear
+    attr(out, "order") <- row.order
+    
+    # TODO: add "data.frame" here too
+    class(out) <- "CrunchDataFrame"
+    return(out)
+}
+
+setOldClass("CrunchDataFrame")
+
+#' @export
+#' @rdname CrunchDataFrame
+dim.CrunchDataFrame <- function (x) {
+    if (is.null(attr(x, "order"))) {
+        # if there is no ordering, the num of rows is the same as dataset
+        n_rows <- nrow(attr(x, "crunchDataset"))
+    } else {
+        # if there is an ordering, the num of rows is the length of the ordering
+        n_rows <- length(attr(x, "order"))
+    }
+    return(c(n_rows, length(ls(x))))
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+names.CrunchDataFrame <- function (x) attr(x, "col_names")
+
+
+
+#' @export
+#' @rdname CrunchDataFrame
+`[.CrunchDataFrame` <- function (x, i, j, drop = TRUE) {
+    j <- grab_col_names(x, j)
+    row_inds <- grab_row_ind(x, i)
+    
+    # grab the columns
+    out <- lapply(j, function(col) {
+        col_data <- x[[col]]
+        if (is.data.frame(col_data)) {
+            return(col_data[row_inds,])
+        } else {
+            return(col_data[row_inds])
+        }
+    })
+    
+    # set names, if j is a character use that, if not grab names from the CDF
+    if (is.character(j)) {
+        names(out) <- j
+    } else {
+        names(out) <- names(x)[j]
+    }
+    
+    if (length(out) == 1 & drop) {
+        # single column, so return vector
+        out <- out[[1]]
+    } else {
+        out <- structure(out, class="data.frame", row.names=seq_along(row_inds))
+    }
+    
+    return(out)
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`[<-.CrunchDataFrame` <- function (x, i, j, value) {
+    j <- grab_col_names(x, j)
+    row_inds <- grab_row_ind(x, i)
+    
+    # TODO: check value length
+    
+    # split value string by the length of row indicators
+    values <- split(value, ceiling(seq_along(value)/length(row_inds)))
+    for (i in seq_along(j)) {
+        col_name <- j[i]
+        x <- set_CDF_var(col_name = col_name, row_inds = row_inds,
+                         cdf = x, value = values[[i]])
+    }
+    return(x)
+}
+
+grab_col_names <- function (x, j) {
```

Comment:
Add some docstrings/comments for these internal functions. Their names aren't clear enough to explain why they exist (i.e. in this case, why you need a function other than "names()").

Also, the snake casing of these function names is a bit inconsistent with the naming conventions used in the package.